p2p/smc_service.py

```diff
@@ -0,0 +1,153 @@
+import asyncio
+import collections
+from typing import (
+    DefaultDict,
+    Dict,
+    List,
+    Optional,
+    Set,
+)
+
+from evm.rlp.headers import (
+    CollationHeader,
+)
+
+from evm.exceptions import (
+    ValidationError,
+)
+from eth_typing import (
+    Address,
+    Hash32,
+)
+from eth_utils import (
+    to_checksum_address,
+)
+from eth_keys import (
+    datatypes,
+)
+
+from p2p.service import (
+    BaseService,
+)
+from p2p.cancel_token import (
+    CancelToken,
+)
+
+from web3 import (
+    Web3,
+)
+
+from sharding.handler.smc_handler import (
+    SMCHandler,
+)
+from sharding.handler.log_handler import (
+    LogHandler,
+)
+from sharding.handler.shard_tracker import (
+    ShardTracker,
+)
+from sharding.contracts.utils.config import (
+    get_sharding_config,
+)
+
+
+class SMCService(BaseService):
+    w3: Web3 = None
+    config = get_sharding_config()
+
+    def __init__(self,
+                 smc_address: Address,
+                 shard_ids: List[int]=None,
+                 private_key: Optional[datatypes.PrivateKey]=None,
+                 token: CancelToken=None) -> None:
+        if self.w3 is None:
```

Comment:
Hrm, class level variable here seems potentially problematic.  Maybe this should be part of the `__init__` params.?